src/metabase/models/dashboard.clj

```diff
@@ -237,6 +237,14 @@
       (doseq [update-card to-update]
         (dashboard-card/update-dashboard-card! update-card (id->current-card (:id update-card)))))))
 
+(defn- remove-invalid-dashcards
+  "Given a list of dashcards, remove any dashcard that link to cards that are either archived or not exist."
+  [dashcards]
+  (let [cards-ids       (remove nil? (map :card_id dashcards))
+        active-card-ids (when-let [card-ids (seq cards-ids)]
+                          (t2/select-pks-set :model/Card :id [:in card-ids] :archived false))]
+    (filter #(or (nil? (:card_id %)) (contains? active-card-ids (:card_id %))) dashcards)))
+
```

Comment:
There is a way to implement this that more closely aligns with the english in the docstring, and is easier to verify that it's correct.
```suggestion
(defn- remove-invalid-dashcards
  "Given a list of dashcards, remove any dashcard that references cards that are either archived or not exist."
  [dashcards]
  (let [card-ids          (set (keep :card_id dashcards))
        active-card-ids   (when-let [card-ids (seq card-ids)]
                            (t2/select-pks-set :model/Card :id [:in card-ids] :archived false))
        inactive-card-ids (set/difference card-ids active-card-ids)]
    (remove #(contains? inactive-card-ids (:card_id %)) dashcards)))
```

Reply:
cool, I've applied it in 84f8fdbabc